is it is expected to have 19 GB of 1.1 mb log files in worksapceStorage

I was running out of diskspace and found that my /Users/manderse/Library/Application Support/Cursor/User/workspaceStorage
contained 29.6 GB.

19.4 GB of that was redhat.java folder with 18442 client.log.<timestamp> files of about 1.1 mb 

whats weird is that its all from about the same sday 2023-09-26 and 2023-09-27.

as if something "bad" happend that day and no cleanup :)

example of what in the log files:

```
  'Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager.getPreferences()" because the return value of "org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin.getPreferencesManager()" is null\n' +
    '\tat org.eclipse.jdt.ls.core.internal.preferences.Preferences.setJavaCompletionFavoriteMembers(Preferences.java:1360)\n' +
    '\tat org.eclipse.jdt.ls.core.internal.preferences.Preferences.createFrom(Preferences.java:1120)\n' +
```

Same thing happening for me, when opening a folder which has never been opened before: after activating java extension pack log gets spammed hard. Can we get at least some kind of fix which stop spamming stacktraces in rate of 10+ per second?

stacktrace:
```
message: říj 30, 2023 4:13:59 ODP. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
  SEVERE: Internal error: java.lang.reflect.InvocationTargetException
  java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:67)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
    at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
    at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)
    at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
    at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
    at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
    at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
    at java.base/java.lang.Thread.run(Unknown Source)
  Caused by: java.lang.reflect.InvocationTargetException
    at jdk.internal.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.base/java.lang.reflect.Method.invoke(Unknown Source)
    at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)
    ... 12 more
  Caused by: java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager.getPreferences()" because the return value of "org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin.getPreferencesManager()" is null
    at org.eclipse.jdt.ls.core.internal.preferences.Preferences.setJavaCompletionFavoriteMembers(Preferences.java:1378)
    at org.eclipse.jdt.ls.core.internal.preferences.Preferences.createFrom(Preferences.java:1126)
    at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.handleInitializationOptions(BaseInitHandler.java:113)
    at org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:105)
    at org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:64)
    at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:284)
    ... 16 more
level: info,
timestamp: 2023-10-30 16:13:59.597
```

Those logs are **extremely** helpful. There is a failure in the Gradle build (from the server logs, `.log`) that would be nice to investigate but I don't think that's the root of the problem. The client logs clearly show the failure happens when we attempt to get a reference to the bundle context for `jdt.ls.core`. After the failure, we seem to get stuck in some `initialize` loop, because the startup was not correctly finished.

Beacuse we're dealing with 2 sets of logs that both give useful info I've tried to reconstruct the failure based on the timestamps :

<details>
<summary>log info</summary>

```
{
  message: "[Trace - 11:54:46 AM] Sending request 'initialize - (0)'.",
  level: 'info',
  timestamp: '2023-11-03 11:54:46.358'
}

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:47.566
!MESSAGE class org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin is started

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:47.735
!MESSAGE Main thread is waiting

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:47.758
!MESSAGE >> initialize

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:47.759
!MESSAGE Initializing Java Language Server 1.29.0.202310260336

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:48.637
!MESSAGE Parent process stopped running, forcing server exit

!ENTRY org.eclipse.jdt.ls.core 1 0 2023-11-03 11:54:48.637
!MESSAGE >> exit

{
  message: 'lis 03, 2023 11:54:48 DOP. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError\r\n' +
    'SEVERE: Internal error: java.lang.reflect.InvocationTargetException\r\n' +
    'java.lang.RuntimeException: java.lang.reflect.InvocationTargetException\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:67)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)\r\n' +
    '\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)\r\n' +
    '\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\r\n' +
    '\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\r\n' +
    '\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\r\n' +
    '\tat java.base/java.lang.Thread.run(Unknown Source)\r\n' +
    'Caused by: java.lang.reflect.InvocationTargetException\r\n' +
    '\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n' +
    '\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n' +
    '\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n' +
    '\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n' +
    '\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)\r\n' +
    '\t... 12 more\r\n' +
    'Caused by: java.lang.NullPointerException: Cannot invoke "org.osgi.framework.BundleContext.getBundle()" because "context" is null\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.handlers.BundleUtils.startBundles(BundleUtils.java:267)\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.handlers.BundleUtils.loadBundles(BundleUtils.java:170)\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:109)\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:64)\r\n' +
    '\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:284)\r\n' +
    '\t... 17 more\r\n' +
    '\r\n',
  level: 'info',
  timestamp: '2023-11-03 11:54:48.743'
}

{
  message: "[Trace - 11:54:48 AM] Received response 'initialize - (0)' in 2385ms. Request failed: Internal error. (-32603).",
  level: 'info',
  timestamp: '2023-11-03 11:54:48.744'
}
{
  message: 'Error data: "java.lang.RuntimeException: java.lang.reflect.InvocationTargetException\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:67)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)\\r\\n\\tat org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$1(ParentProcessWatcher.java:144)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)\\r\\n\\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)\\r\\n\\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\\r\\n\\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\\r\\n\\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\\r\\n\\tat java.base/java.lang.Thread.run(Unknown Source)\\r\\nCaused by: java.lang.reflect.InvocationTargetException\\r\\n\\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\\r\\n\\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\\r\\n\\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\\r\\n\\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\\r\\n\\tat org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$recursiveFindRpcMethods$0(GenericEndpoint.java:65)\\r\\n\\t... 12 more\\r\\nCaused by: java.lang.NullPointerException: Cannot invoke \\"org.osgi.framework.BundleContext.getBundle()\\" because \\"context\\" is null\\r\\n\\tat org.eclipse.jdt.ls.core.internal.handlers.BundleUtils.startBundles(BundleUtils.java:267)\\r\\n\\tat org.eclipse.jdt.ls.core.internal.handlers.BundleUtils.loadBundles(BundleUtils.java:170)\\r\\n\\tat org.eclipse.jdt.ls.core.internal.handlers.InitHandler.handleInitializationOptions(InitHandler.java:109)\\r\\n\\tat org.eclipse.jdt.ls.core.internal.handlers.BaseInitHandler.initialize(BaseInitHandler.java:64)\\r\\n\\tat org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.initialize(JDTLanguageServer.java:284)\\r\\n\\t... 17 more\\r\\n"\n' +
    '\n',
  level: 'info',
  timestamp: '2023-11-03 11:54:48.744'
}
{
  message: 'Failed to initialize Language Support for Java due to Error: Internal error.',
  level: 'error',
  timestamp: '2023-11-03 11:54:48.744'
}
{
  message: "[Trace - 11:54:48 AM] Sending request 'initialize - (1)'.",
  level: 'info',
  timestamp: '2023-11-03 11:54:48.745'
}

...
...

loop begins
```
</details>

At some point the language server detects that the parent process (the client with PID 18392) stopped running. After that it begins shutting down while initialization is still occuring. This causes failures on the client side, which ultimately causes the client to re-attempt to initialize.

The good news is, we have code in vscode-java that can detect this and should be able to prevent the initialization loop that causes the massive size in storage. I don't know what is causing the parent process watcher to think the client is gone.

--- a/src/extension.ts
+++ b/src/extension.ts
@@ -29,7 +29,7 @@ import { snippetCompletionProvider } from './snippetCompletionProvider';
 import { JavaClassEditorProvider } from './javaClassEditor';
 import { StandardLanguageClient } from './standardLanguageClient';
 import { SyntaxLanguageClient } from './syntaxLanguageClient';
-import { convertToGlob, deleteDirectory, ensureExists, getBuildFilePatterns, getExclusionBlob, getInclusionPatternsFromNegatedExclusion, getJavaConfig, getJavaConfiguration, hasBuildToolConflicts } from './utils';
+import { convertToGlob, deleteClientLog, deleteDirectory, ensureExists, getBuildFilePatterns, getExclusionBlob, getInclusionPatternsFromNegatedExclusion, getJavaConfig, getJavaConfiguration, hasBuildToolConflicts } from './utils';
 import glob = require('glob');
 import { Telemetry } from './telemetry';
 import { getMessage } from './errorUtils';
@@ -110,7 +110,12 @@ export async function activate(context: ExtensionContext): Promise<ExtensionAPI>
 	if (!storagePath) {
 		storagePath = getTempWorkspace();
 	}
+	const workspacePath = path.resolve(`${storagePath}/jdt_ws`);
 	clientLogFile = path.join(storagePath, 'client.log');
+	const cleanWorkspaceExists = fs.existsSync(path.join(workspacePath, cleanWorkspaceFileName));
+	if (cleanWorkspaceExists) {
+		deleteClientLog(storagePath);
+	}
 	initializeLogFile(clientLogFile);
 
 	const telemetryService: Promise<TelemetryService> = Telemetry.startTelemetry(context);
@@ -137,7 +142,6 @@ export async function activate(context: ExtensionContext): Promise<ExtensionAPI>
 	}).then(async (requirements) => {
 		const triggerFiles = await getTriggerFiles();
 		return new Promise<ExtensionAPI>(async (resolve) => {
-			const workspacePath = path.resolve(`${storagePath}/jdt_ws`);
 			const syntaxServerWorkspacePath = path.resolve(`${storagePath}/ss_ws`);
 
 			let serverMode = getJavaServerMode();
@@ -254,7 +258,7 @@ export async function activate(context: ExtensionContext): Promise<ExtensionAPI>
 				errorHandler: new ClientErrorHandler(extensionName),
 				initializationFailedHandler: error => {
 					logger.error(`Failed to initialize ${extensionName} due to ${error && error.toString()}`);
-					if (error.toString().includes('Connection') && error.toString().includes('disposed')) {
+					if ((error.toString().includes('Connection')  && error.toString().includes('disposed')) || error.toString().includes('Internal error')) {
 						if (!initFailureReported) {
 							apiManager.fireTraceEvent({
 								name: "java.client.error.initialization",
@@ -315,7 +319,6 @@ export async function activate(context: ExtensionContext): Promise<ExtensionAPI>
 				}
 			}));
 
-			const cleanWorkspaceExists = fs.existsSync(path.join(workspacePath, cleanWorkspaceFileName));
 			if (cleanWorkspaceExists) {
 				const data = {};
 				try {
@@ -1049,3 +1052,5 @@ function registerRestartJavaLanguageServerCommand(context: ExtensionContext) {
 		}
 	}));
 }
+
+

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -35,6 +35,19 @@ export function deleteDirectory(dir) {
 	}
 }
 
+export function deleteClientLog(dir) {
+	if (fs.existsSync(dir)) {
+		fs.readdirSync(dir).forEach((child) => {
+			if (child.startsWith('client.log') || child.endsWith('audit.json')) {
+				const entry = path.join(dir, child);
+				if (!fs.lstatSync(entry).isDirectory()) {
+					fs.unlinkSync(entry);
+				}
+			}
+		});
+	}
+}
+
 export function getTimestamp(file) {
 	if (!fs.existsSync(file)) {
 		return -1;

